proof/refine/ARM/Ipc_R.thy

```diff
@@ -3560,160 +3560,622 @@ lemma cteInsert_ct'[wp]:
-                          st_tcb_at_refs_of_rev' pred_tcb_at'
-                          conj_comms fun_upd_def[symmetric]
-               split del: if_split)
-    apply (frule pred_tcb_at')
-    apply (drule simple_st_tcb_at_state_refs_ofD' st_tcb_at_state_refs_ofD')+
-    apply (clarsimp simp: valid_pspace'_splits)
-    apply (subst fun_upd_idem[where x=t])
-     apply (clarsimp split: if_split)
-     apply (rule conjI, clarsimp simp: obj_at'_def projectKOs)
-     apply (drule bound_tcb_at_state_refs_ofD')
-     apply (fastforce simp: tcb_bound_refs'_def)
-    apply (subgoal_tac "ex_nonz_cap_to' a s")
-     prefer 2
-     apply (clarsimp elim!: if_live_state_refsE)
-    apply clarsimp
-    apply (rule conjI)
-     apply (drule bound_tcb_at_state_refs_ofD')
-     apply (fastforce simp: tcb_bound_refs'_def set_eq_subset)
-    apply (clarsimp simp: conj_ac)
-    apply (rule conjI, clarsimp simp: idle'_only_sc_refs)
-    apply (rule conjI, clarsimp simp: global'_no_ex_cap)
-    apply (rule conjI)
-     apply (rule impI)
-     apply (frule(1) ct_not_in_epQueue, clarsimp, clarsimp)
-     apply (clarsimp)
-    apply (simp add: ep_redux_simps')
-    apply (rule conjI, clarsimp split: if_split)
-     apply (rule conjI, fastforce simp: tcb_bound_refs'_def set_eq_subset)
-     apply (clarsimp, erule delta_sym_refs;
-            solves\<open>auto simp: symreftype_inverse' tcb_bound_refs'_def split: if_split_asm\<close>)
-    apply (solves\<open>clarsimp split: list.splits\<close>)
+    apply (case_tac list; simp)
+    apply (wpsimp wp: possibleSwitchTo_invs' si_invs'_helper replyUnlink_invs'
+                      replyUnlink_st_tcb_at' threadGet_wp setThreadState_st_tcb
+                      setThreadState_Running_invs' hoare_weak_lift_imp)
+        apply (rule_tac Q="\<lambda>_ s. invs' s \<and> st_tcb_at' active' t s \<and> sch_act_not t s \<and>
+                                 ex_nonz_cap_to' t s \<and> ex_nonz_cap_to' a s \<and> t \<noteq> a \<and>
+                                 (\<forall>x. replyObject recvState = Some x \<longrightarrow> ex_nonz_cap_to' x s) \<and>
+                                 (cd \<and> bound_sc_tcb_at' ((=) None) a s \<longrightarrow>
+                                   (\<exists>sc. bound_sc_tcb_at' ((=) sc) t s \<and>
+                                         ex_nonz_cap_to' (the sc) s))"
+                     in hoare_strengthen_post[rotated])
+         apply (fastforce simp: obj_at'_def projectKO_eq projectKO_tcb
+                                pred_tcb_at'_def invs'_def valid_state'_def
+                         dest!: global'_no_ex_cap)
+        apply (wpsimp simp: invs'_def valid_state'_def valid_pspace'_def comp_def sym_refs_asrt_def
+                        wp: hoare_vcg_all_lift hoare_vcg_ex_lift hoare_vcg_imp_lift' gts_wp')+
+    apply (intro conjI; clarsimp?)
+        apply (auto simp: obj_at'_def projectKO_eq projectKO_ep valid_obj'_def valid_ep'_def
```

Comment:
Auto here too

Reply:
again, I think this one is ok, too, since it’s limited to one subgoal. You can do `subgoal`, too, of course, but I don’t think there is particular advantage in choosing that over `[1]` here. I might be wrong, though...

Reply:
Changed to force since fastforce didn't work